pkg/northbound/gnmi/subscribe.go

```diff
@@ -48,78 +49,89 @@ func (s *Server) Subscribe(stream gnmi.GNMI_SubscribeServer) error {
 			return err
 		}
 
-		subscribe := in.GetSubscribe()
-		mode := subscribe.Mode
+		var mode gnmi.SubscriptionList_Mode
+
+		if in.GetPoll() != nil {
+			mode = gnmi.SubscriptionList_POLL
+		} else {
+			subscribe = in.GetSubscribe()
+			mode = subscribe.Mode
+		}
 		stopped := make(chan struct{})
-		//If the subscription mode is ONCE we immediately start a routine to collect the data
-		if mode == gnmi.SubscriptionList_ONCE {
+		//If the subscription mode is ONCE or POLL we immediately start a routine to collect the data
+		if mode != gnmi.SubscriptionList_STREAM {
 			go collector(updateChan, subscribe)
 		}
-		//TODO for POLL type spawn a routine that periodically checks for updates
+
 		//This generate a subscribe response for one or more updates on the channel.
 		// for Subscription_once messages also also closes the channel.
-		go func() {
-			for update := range updateChan {
-				updateArray := make([]*gnmi.Update, 0)
-				updateArray = append(updateArray, update)
-				notification := &gnmi.Notification{
-					Timestamp: time.Now().Unix(),
-					Update:    updateArray,
-				}
-				responseUpdate := &gnmi.SubscribeResponse_Update{
-					Update: notification,
-				}
-				response := &gnmi.SubscribeResponse{
-					Response: responseUpdate,
-				}
-				sendResponse(response, stream)
-				//For stream and Poll we also send a Sync Response
-				//TODO make sure that for stream sending this every time adheres to spec.
-				// see section #3.5.1.4 of gnmi-specification.md
-				if mode != gnmi.SubscriptionList_ONCE {
-					responseSync := &gnmi.SubscribeResponse_SyncResponse{
-						SyncResponse: true,
-					}
-					response = &gnmi.SubscribeResponse{
-						Response: responseSync,
-					}
-					sendResponse(response, stream)
-				} else {
-					//If the subscription mode is ONCE we read from the channel, build a response and issue it
-					stopped <- struct{}{}
-				}
-			}
-		}()
+		go listenForUpdates(updateChan, stream, mode, stopped)
```

Comment:
yes, that is the internal channel where we send the internal update, we use it to stream the gnmi.update we get from the manager in case of ONCE and POLL or the gnmi.update generated by the event the set issues. so we use this channel one time for the ONCE too. If you look at the stopped channel, that's how I make sure to exit the for loop and return a nil error

Reply:
got it